Update Frontier machine and compiler configuration to fix issues with crayclanggpu and amdclanggpu

grnydawn opened the pull request:
The update includes:
* Changed Core from 24.07 to 24.00.
* Updated CMake from 3.27.9 to 3.21.3.
* Restored --allow-multiple-definition in CMAKE_EXE_LINKER_FLAGS.
* Removed the craype-accel-amd-gfx90a module load for amdclanggpu_frontier.
* Added two MOSART source files to NOOPT_FILES as a workaround to prevent a crayclanggpu build error.

Fixes #6750, Fixes #6755

Note: This PR disables OpenMP offload and OpenACC for the AMD compiler on Frontier.

Note: Optimization is disabled for the following MOSART source files:
* mosart/src/wrm/WRM_subw_IO_mod.F90
* mosart/src/riverroute/RtmMod.F90

[BFB] No baseline for Frontier yet.

grnydawn commented:
Testing is ongoing, and the results will be posted here once it finishes. So far, the test outcomes are not worse than the previous test results; in some cases, they are better.

Please invite anyone to review this PR if you think it is necessary.

rljacob commented:
Are you trying to get scream cases to build and run with this PR?

dqwu commented:
> Are you trying to get scream cases to build and run with this PR?

This PR tries to fix some known build issue for machine frontier. No changes to machine frontier-scream-gpu.
I think it has nothing to do with scream cases so the title might needs to be updated.

rljacob commented:
We want to abolish frontier-scream-gpu and be able to run scream cases with the frontier machine entry.  Perhaps that will wait for another PR.

dqwu commented:
@grnydawn If this PR can fix issue 6750, 6755, 6764, you should probably add fixes #xxxx in the description.

grnydawn commented:
> We want to abolish frontier-scream-gpu and be able to run scream cases with the frontier machine entry. Perhaps that will wait for another PR.

@rljacob , Sure, I will work on removing frontier-scream-gpu with the Scream group after this PR.

sarats commented:
> Changed Core from 24.07 to 24.00.
> Updated CMake from 3.27.9 to 3.21.3.

Just curious, why the downgrade in versions? esp. CMake?

sarats commented:
Also FYI (my comment in the file about -Wl,--allow-shlib-undefined). I would vote to leave it in to prevent similar issues, it's typically harmless.
```
# -Wl,--allow-shlib-undefined was added to address rocm 5.4.3 Fortran linker issue:
# /opt/rocm-5.4.3/lib/libhsa-runtime64.so.1: undefined reference to `std::condition_variable::wait(std::unique_lock<std::mutex>&)@GLIBCXX_3.4.30'
# AMD started building with GCC 12.2.0, which brings in a GLIBCXX symbol that isn't in CCE's default GCC toolchain.
```

dqwu commented:
> Also FYI (my comment in the file about -Wl,--allow-shlib-undefined). I would vote to leave it in to prevent similar issues, it's typically harmless.
> 
> ```
> # -Wl,--allow-shlib-undefined was added to address rocm 5.4.3 Fortran linker issue:
> # /opt/rocm-5.4.3/lib/libhsa-runtime64.so.1: undefined reference to `std::condition_variable::wait(std::unique_lock<std::mutex>&)@GLIBCXX_3.4.30'
> # AMD started building with GCC 12.2.0, which brings in a GLIBCXX symbol that isn't in CCE's default GCC toolchain.
> ```

This one can be restored if your prefer to do so, though it looks optional so far.

grnydawn commented:
> > Changed Core from 24.07 to 24.00.
> > Updated CMake from 3.27.9 to 3.21.3.
> 
> Just curious, why the downgrade in versions? esp. CMake?

@sarats , Some combinations of particular versions of CMake and Core mode work with the Scorpio build with threading, while others do not. Please see E3SM Issue #6750 for the discussions.

grnydawn commented:
> > Are you trying to get scream cases to build and run with this PR?
> 
> This PR tries to fix some known build issues for machine frontier. No changes to machine frontier-scream-gpu. I think it has nothing to do with scream cases so the title might needs to be updated.

@dqwu ,

grnydawn commented:
> @grnydawn If this PR can fix issue 6750 and 6755 (6764 might only be partially fixed for some cases), you should probably add fixes #xxxx in the description.

@dqwu , I added notes for fixing issues #6750 and #6755. Issue #6764 has not been verified yet and will be updated if it is also fixed.

dqwu commented:
> > @grnydawn If this PR can fix issue 6750 and 6755 (6764 might only be partially fixed for some cases), you should probably add fixes #xxxx in the description.
> 
> @dqwu , I added notes for fixing issues #6750 and #6755. Issue #6764 has not been verified yet and will be updated if it is also fixed.

FYI, in GitHub, the correct format to auto-close multiple issues is:
Fixes: #xxxx, #yyyy
The key here is to separate the issue numbers with a comma, without any extra spaces between them.
Please update the description accordingly, otherwise issue 6755 will not be auto-closed.

grnydawn commented:
> > > @grnydawn If this PR can fix issue 6750 and 6755 (6764 might only be partially fixed for some cases), you should probably add fixes #xxxx in the description.
> > 
> > 
> > @dqwu , I added notes for fixing issues #6750 and #6755. Issue #6764 has not been verified yet and will be updated if it is also fixed.
> 
> @grnydawn FYI, in GitHub, the correct format to auto-close multiple issues is: Fixes: #xxxx, #yyyy
> 
> The key here is to separate the issue numbers with a comma, without any extra spaces between them. Please update the description accordingly, otherwise issue 6755 will not be auto-closed.

@dqwu , I updated my PR description. Thanks!

philipwjones reviewed (approved):
Changes are fine with me
